### PR TITLE
[10.0][ADD] - Survey five stars question type

### PR DESCRIPTION
**Pull request by sbejaoui**
first commit for Survey five stars question type

### CODE REVIEW

File: survey_five_stars_question_type/README.rst
Comment (benwillig) [inline survey_five_stars_question_type/README.rst]: s/project_repo/survey

File: survey_five_stars_question_type/README.rst
Comment (benwillig) [inline survey_five_stars_question_type/README.rst]: s/repo_id/200
s/branch/10.0

File: survey_five_stars_question_type/__init__.py
Comment (benwillig) [inline survey_five_stars_question_type/__init__.py]: Missing blank line at the end of file

File: survey_five_stars_question_type/models/survey_user_input_line.py
Comment (benwillig) [inline survey_five_stars_question_type/models/survey_user_input_line.py]: s/SurveyUser_input_line/SurveyUserInputLine

File: survey_five_stars_question_type/models/survey_user_input_line.py
Comment (benwillig) [inline survey_five_stars_question_type/models/survey_user_input_line.py]: Shouldn't we add a python constrains which check that the answer is between 1 and 5?

File: survey_five_stars_question_type/tests/test_survey.py
Comment (benwillig) [inline survey_five_stars_question_type/tests/test_survey.py]: Shouldn't we check that there is no error when providing a valid value?

File: survey_five_stars_question_type/views/survey_template.xml
Comment (benwillig) [inline survey_five_stars_question_type/views/survey_template.xml]: data tag is deprecated

**Review by benwillig — CHANGES_REQUESTED** [review]
Some minor comments.

Could you split the readme.rst in separate files like [here](https://github.com/OCA/mis-builder/tree/10.0/mis_builder/readme). The readme file will be generated automatically after the PR is merged.

File: survey_five_stars_question_type/__manifest__.py
Comment (benwillig) [inline survey_five_stars_question_type/__manifest__.py]: The website should be the OCA repository url.

**Comment by pedrobaeza** [comment]
Please avoid plurals on the name, and following usual conventions, the name should be `survey_question_type_five_star`.

**Comment by pedrobaeza** [comment]
Check also Travis status

File: survey_five_stars_question_type/models/survey_user_input_line.py
Comment (sbejaoui) [reply survey_five_stars_question_type/models/survey_user_input_line.py]: Already done at `models/survey_question.py: validate_star_rate`

Comment (benwillig) [reply survey_five_stars_question_type/models/survey_user_input_line.py]: I'm talking about a python model constrain, because if I do a simple ```write``` with an invalid value, there will be no error.

Comment (sbejaoui) [reply survey_five_stars_question_type/models/survey_user_input_line.py]: Yes I agree, a simple `write` bypass the constraint, we may have a question type matrix and as answer we put text.
But it's done like that in survey module, the check is done only at user submit.
A method `save_{question_type}` call `validate_{question_type}`.

Comment (benwillig) [reply survey_five_stars_question_type/models/survey_user_input_line.py]: Odoo doesn't use pylint/flake8 but we do :smile:
We might need more opinion about that but I would like to have a constrain for that. I would say, it isn't because Odoo did it wrong that we should follow the same path ^^.

**Review by benwillig — CHANGES_REQUESTED** [review]
After some functionnal testing, I found some points which need to be fixed

1) When answering to a survey with the new question type, if you don't click on any stars and you submit the form. There will be an error. This error is raised because the value for the answer is blank and isn't between 1 and 5. So I think,we should allow people to mark the answer without any star, with a 0 value then.

2) After answering the survey, you can check the answer you gave. The new question type is not handled in this view.

3) If you want to check the survey results with the "View results" button available in the survey form view. The new question type is not handled here so we can't have any statistics.

File: survey_question_type_five_star/static/src/js/survey.js
Comment (benwillig) [inline survey_question_type_five_star/static/src/js/survey.js:15]: This call is already done in the base module. Can't we just add a onchange listener on the input? Then when the base module updates inputs value, you can determine which star should be highlighted. I think we could use the ```Widget``` class, this widget would take a DOM element/JS selector in parameter. And the target object would be the HTML structure of the "Five star rate" representation. When the widget is instanciated, it would check the current value of the input to highlight stars, and it would add an onchange listener to handle the possible changes. The widget should be readonly for the result view and editable when answering.

File: survey_question_type_five_star/static/src/js/survey.js
Comment (benwillig) [inline survey_question_type_five_star/static/src/js/survey.js]: We have two different logics to highlight stars, so I think a widget would be better here.

**Review by benwillig — CHANGES_REQUESTED** [review]

File: survey_question_type_five_star/models/survey_survey.py
Comment (tarteo) [inline survey_question_type_five_star/models/survey_survey.py:37]: If this is stored in a variable (like sum_all_inputs) and in the 2 cases it's more efficient. But the gain is minimal so it's not blocking :P

File: survey_question_type_five_star/static/src/js/survey.js
Comment (tarteo) [inline survey_question_type_five_star/static/src/js/survey.js]: New line

**Review by tarteo — CHANGES_REQUESTED** [review]
Cool module! Code looks good, I just have a few minor comments. 
When I make a question mandatory, I still can skip it. 
It would be nice be to be able to select the amount of stars for a question.

**Comment by sbidoul** [comment]
Hi all, thanks for the review. We have limited budget to finalize this.

I see one important issue: "When I make a question mandatory, I still can skip it.". @sbejaoui is that easy to fix?

@tarteo Making the number of stars configurable looks too complex for our budget. May be we can rename the module to survey_question_type_star, and leave the configurability for later.

**Comment by sbidoul** [comment]
[updated] module name in previous comment

**Comment by sbidoul** [comment]
@benwillig I understand your comments are nice to have refactorings. Are they blocking for you?

**Comment by sbejaoui** [comment]
@sbidoul , @tarteo ,
For the mandatory option, I think we should decide if we consider 0 as a valid answer or not.
When user skip a question and any of the five stars is highlighted a 0 value is sent.
In this case the mandatory option make no sens.
So I propose to consider valid answer for mandatory question from 1 to 5.

**Review by tarteo — APPROVED** [review]
@sbejaoui @sbidoul I tested it again and it works perfectly. Thanks. Do you have budget to fix the jslint errors?

**Comment by sbejaoui** [comment]
@tarteo , I pushed a fix for the jslint

**Review by rafaelbn — APPROVED** [review]

**Comment by sbejaoui** [comment]
Hi @rafaelbn ,

I believe this can be merged

**Pull request closed without merging**
